Adds target support for linux over SSH in ec2

burtlo opened the pull request:
## LINUX - Centos 6 Running in AWS

Running the command

```bash
$ bin/ohai --target ssh://centos@IPADDRESS -i ~/.ssh/ohai.pem
```

### Plugins

Need to fix the following plugin discrepancies.

- [x] `languages`
- [x] `uptime`
- [x] `counters` is empty
- [x] `network` is empty
- [x] `virtualization` missing `"xen": "guest"`
- [x] `packages`
- [x] `sysconf` is missing
- [x] `filesystem` 
- [x] `dmi` is missing `"dmidecode_version": "2.12"`
- [x] `keys/ssh`
- [x] `block_device`
- [x] `etc` is pulling in some local entries (e.g. `_appleevents` appearing in a centos target)
- [x] `root_group` is collecting data locally
- [x] `current_user` is collecting data locally
- [x] `time/timezone` is collecting data locally
- [x] `ec2` is missing and that is because the `lib/mixin/ec2_metadata` assumes the code is run locally and performs an HTTP from the local instance to the IP only available remotely. It seems like it could be replaced with some rudimentary curling and relying on exit status.
- [x] `cloud` is null because it is dependent on ec2
- [x] `json?` attribute missing something `"{\n  \"devpayProductCodes\" : null....region\" : \"us-east-1\"\n}",
- [x] `virtualization/systems/docker` is reporting 'host' when normally it is absent
- [x] `virtualization/systems/openstack` is reporting 'host' when normally it is absent
- [x] `virtualization/system` is reporting 'openstack' when normally it is 'xen'
- [x] `virtualization/role` is reporting 'host' when normally it is 'guest'
- [ ] `docker` key is present and empty when it is usually absent from the results (this is minor)
- [x] `memory/directmap` is found when executed remotely but not locally - turns out that is just not in the version of ohai running on the system.
- [x] `chef_packages` is collecting data locally and using `require 'chef/version'` to get to that information causing an exception to be raised (this currently pauses the execution)

### Performance

*Running Locally* with Ohai 14.8.10 (2 seconds)
```
real	0m1.492s
user	0m0.523s
sys	0m0.257s
```
*Running remotely* with this Ohai (2 minutes)
```
2.35s user 0.50s system 2% cpu 2:01.49 total
```

lynnvfrank committed:
WIP adds remote support to ohai

- added a limited flag support to the cli
- added train gem as a dependency
- `Ohai::System` reads the config and creates a backend with train and creates a connection
- `Ohai::Runner` then inserts the connection into each ohai plugin
- Rewrote the module that brings in `shell_out` support to use the connection.
- Rewrote the module that checks the `which` on the system to use the connection’s file to get the mode to determine if it is executable

Limitations:
- Now if a plugin uses something besides `shell_out` it slips by this implementation.
- All uses of Ruby’s `File` class will also need to be replaced in every plugin.

Signed-off-by: Franklin Webber <franklin.webber@gmail.com>

aaronlippold commented:
Thoughts / What makes this complete ( perhaps leveraging code from InSpec / Train )? 
- Enable a `lazy` or `cache` approach by default so we don't have huge network lag due to train envoking ohai over and over to just get the same data.
  - allow the user to say - yes - please refresh the data cuz I want you to.
- Ensure support for all transports ( I assume train already generalized this via a mixing )
  - Core transport support would be:  aws, docker, kubernetes, oc, winrm, ssh
- Ensure auth method support for windows and linux ( aka ssh (done?) and winrm certs, etc )
- Examine if we need to add other Train classes - such as Directory - and if this can be MixedIn
- Evaluate the extent of `shell_out` in the Train codebase and if a default `shell_out` can be attached to all plugin objects rather than having to recode non-compliant plugins.

aaronlippold commented:
Would the proper order of operations be - once this is in the train core - that the `ohai` resource tries to use this first, then if needed, fail back to local client install else throw InSpec::Exception::FailResource.

aaronlippold commented:
https://chefcommunity.slack.com/files/U1ZNUCKDX/FHHD1UELE/2019-03-30_01-33-11.mp4

burtlo commented:
@aaronlippold wrote: 
> Would the proper order of operations be - once this is in the train core - that the ohai resource tries to use this first, then if needed, fail back to local client install else throw InSpec::Exception::FailResource.

In my head, you would want to to compose a solution with `ohai` that uses `train`:

`ohai` would never talk directly to the OS and instead talk to `train`. This work here would make it so `ohai` could work in that way. With that in place you could perform a remote scan of attributes with a train-powered `ohai` then run the scan with your inspec profie. Then combine the results. [Small, sharp tools](https://brandur.org/small-sharp-tools)

From what I'm gathering from you, the hope is to have `train` include `ohai`. So that from within `train` you can ask it to perform an `ohai` scan and capture all that data  for reporting back to anything using `train`.

From my perspective it feels like there could be quite a bit of value in having train be able to be so thorough in its ability to give so much data back about the platform. But I thought I remember reading from other issues that this is something that most of the people with commit bits would not be interested in seeing.

Now, with that said, I think there is a lot that could be gained from completing this work and then taking a look at the state of things. Because I have a sneaking suspicion that quite a bit of work that would be added here to make sure `ohai` never touches the OS would benefit `train` greatly by creating a more robust interface. One example:

One of the helpers in `ohai` provides something that wants to insure that a file is found on the path and is executable. `train` gives you back the file mode on unix (and I guess perhaps a permission set on windows) but it doesn't give you a lot of nice convenience methods like `connection.file('/usr/bin/git`).executable?` which would take into account the current user (and their groups) and the mode on the file.

burtlo commented:
> Enable a lazy or cache approach by default so we don't have huge network lag due to train envoking ohai over and over to just get the same data.

Right now it is performing a bunch of commands on the same connection. I don't know how much more caching it can get, but I'll examine the exact details of what is going on.

> allow the user to say - yes - please refresh the data cuz I want you to.

Is this a suggestion for `ohai` or another tool that would want to keep polling for new data?

>Ensure support for all transports ( I assume train already generalized this via a mixing )
>        Core transport support would be: aws, docker, kubernetes, oc, winrm, ssh

When I sit back down again I will probably just lift the code straight from inspec that parses the target string.

> Ensure auth method support for windows and linux ( aka ssh (done?) and winrm certs, etc )

I definitely left out password support. This also needs a `--keyfile` flag to provide a key. I hard-coded it for now.

> Examine if we need to add other Train classes - such as Directory - and if this can be MixedIn

I commented out the existing modules that mixed in helpers. In a real implementation I would put that code into the mixins.

> Evaluate the extent of shell_out in the Train codebase and if a default shell_out can be attached to all plugin objects rather than having to recode non-compliant plugins.

That is currently mostly true. Examining every plugin is still required to ensure there is no use of Ruby's `File`, `Dir` and other ways in which it fails to use the backend connection.

burtlo commented:
I had some dinner and thought more on your idea. I see what you are hoping train will become. In some ways it makes a lot of sense. A lot of the resources that currently perform operations solely to extract data would be better codified in something like an Ohai plugin structure. I could see that data being available and then other resources deferring to that location in train to retrieve it. train would invoke the ohai plugin like work to retrieve it and then ...

So sure, I see how it possibly works. I think this could work well for resources where the extraction of the data results in the entire set of data and the resource could then scope down on what was required. This would make it easy for a plugin to get everything, store it, and then give it back.

I try to imagine that interface. Like would I write a lower-level plugin that extracts the data on the platform and then write a higher level resources that acts as an interface to the test interface. I don't think its a bad abstraction at all. But I can't quite picture it in the current context.

lynnvfrank committed:
Adds the keyfile param to the cli

Signed-off-by: Franklin Webber <franklin.webber@gmail.com>

aaronlippold commented:
Well I think we've made some great progress so far. I'm not sure that trying to imagine the entirety of the solution is necessary on the first generation, given that we are trying to just get the core capability in place I'm sure we can iterate and grow it overtime. I think if we were to implement the course set of data that was outlined in the other PR's around the Target metadata geared more towards Target inventory, such as IP address host name Mac address, etc using the embedded ahi train approach and we can grow it as we get better about knowing what the proper abstraction and interface levels are.

aaronlippold commented:
I've seen many issue and pull requests over the years where they would like to expand the reporters the resources etc especially around the area so virtualization containers etc where the context of what target you're on and being able to identify it as a container or a virtual machine or a physical host would be necessary for generalized automation. I think this goes towards solving this problem. I also think that one of the major things that are validation platform has ignored is being clear about a results set source. I shouldn't have to augment the report to know some identifying information about what Target I just ran my profile on. We have the library, we can rebuild it we have the technology :-) etc etc it's already there we just need to integrate them. Lol

lynnvfrank committed:
Provide alts for system specific operations

Give plugins a replacement for file exist read and readlines.
Give plugins a replacment for Dir
Train returns a result with exit_status and not exitstatus

Signed-off-by: Franklin Webber <franklin.webber@gmail.com>

lynnvfrank committed:
Adds error catching in development

Adds password and keyfile support
Adds lots of pry points for catching errors
Signed-off-by: Franklin Webber <franklin.webber@gmail.com>

aaronlippold commented:
WRT: `Is this a suggestion for ohai or another tool that would want to keep polling for new data?`

BLUF: be able to ask train and or `ohai` to rerun the ohai data collection and have a new object from the one that was created in the original train connection to the target. For good reasons or bad, I would guess users would want it.

LONGER:

I would guess the core use-case would be just one poll of the target `ohai` information as we are running one `validation` event via the profile. However, I cannot assume that at some point, for some reason, a user may want to colour outside the lines, and just plain ask for the data set to be refreshed on the target for whatever reason. ( I know it would be generally bad practice as that would likely mean they were changing the target in some way in the middle of the tests but I can't judge that. I would just guess that at some point a user would say, can you add the ability to refresh ... and why not just put it there from the beginning and not have to deal with it down the line. Testing would be easy given that `/proc/` I assume changes all the time ")

lynnvfrank committed:
Add more replacements for File.open

File.open to file_open
Replaced some remaining File.xxxx methods

Signed-off-by: Franklin Webber <franklin.webber@gmail.com>

lynnvfrank committed:
Fixes the file_open helper.

It needed to read the content and put it in the StringIO

Signed-off-by: Franklin Webber <franklin.webber@gmail.com>

lynnvfrank committed:
Provides an alternate way to get the modules

Signed-off-by: Franklin Webber <franklin.webber@gmail.com>

burtlo commented:
## LINUX - Centos 6 Running in AWS

Running the command

```bash
$ bin/ohai --target ssh://centos@IPADDRESS -i ~/.ssh/ohai.pem keys/ssh
```

### Plugins

Need to fix the following plugin discrepancies.

- [x] `languages`
- [x] `uptime`
- [x] `counters` is empty
- [x] `network` is empty
- [x] `virtualization` missing `"xen": "guest"`
- [x] `packages`
- [x] `sysconf` is missing
- [x] `filesystem` 
- [x] `dmi` is missing `"dmidecode_version": "2.12"`
- [x] `keys/ssh`
- [x] `block_device`
- [x] `etc` is pulling in some local entries (e.g. `_appleevents` appearing in a centos target)
- [x] `root_group` is collecting data locally
- [x] `current_user` is collecting data locally
- [x] `time/timezone` is collecting data locally
- [x] `ec2` is missing and that is because the `lib/mixin/ec2_metadata` assumes the code is run locally and performs an HTTP from the local instance to the IP only available remotely. It seems like it could be replaced with some rudimentary curling and relying on exit status.
- [x] `cloud` is null because it is dependent on ec2
- [x] `json?` attribute missing something `"{\n  \"devpayProductCodes\" : null....region\" : \"us-east-1\"\n}",
- [x] `virtualization/systems/docker` is reporting 'host' when normally it is absent
- [x] `virtualization/systems/openstack` is reporting 'host' when normally it is absent
- [x] `virtualization/system` is reporting 'openstack' when normally it is 'xen'
- [x] `virtualization/role` is reporting 'host' when normally it is 'guest'
- [ ] `docker` key is present and empty when it is usually absent from the results (this is minor)
- [x] `memory/directmap` is found when executed remotely but not locally - turns out that is just not in the version of ohai running on the system.
- [x] `chef_packages` is collecting data locally and using `require 'chef/version'` to get to that information causing an exception to be raised (this currently pauses the execution)

### Performance

*Running Locally* with Ohai 14.8.10 (2 seconds)
```
real	0m1.492s
user	0m0.523s
sys	0m0.257s
```
*Running remotely* with this Ohai (2 minutes)
```
2.35s user 0.50s system 2% cpu 2:01.49 total
```

lynnvfrank committed:
Fixes the which helper

I don't think ENV PATH returns anything but now it works at least to
grab the stdout. The ip command will not work without the fullpath
so I replaced all instances of the naked ip with the which('ip').

Signed-off-by: Franklin Webber <franklin.webber@gmail.com>

lynnvfrank committed:
Commenting out all breakpoints
Signed-off-by: Franklin Webber <franklin.webber@gmail.com>

lynnvfrank committed:
Adds which to make the command work

Signed-off-by: Franklin Webber <franklin.webber@gmail.com>

lynnvfrank committed:
Replaces hardcoded lsmod to use which helper
Signed-off-by: Franklin Webber <franklin.webber@gmail.com>

lynnvfrank committed:
Fixes the file_open helper and updates ssh

ssh was using a lot of IO.read that doesn't work
Signed-off-by: Franklin Webber <franklin.webber@gmail.com>

lynnvfrank committed:
Comment out all breakpoints on failure

Signed-off-by: Franklin Webber <franklin.webber@gmail.com>

burtlo commented:
## darwin

Running

```bash
$ bin/ohai
```

- [ ] `languages/java`

lynnvfrank committed:
Adds note about what is poor design

lynnvfrank committed:
Adds better collect_os and is executable

the is_executable is still poor. It is now checkin for another number
but not much more and other executable things will slip by

Signed-off-by: Franklin Webber <franklin.webber@gmail.com>

lynnvfrank committed:
fixes java for darwin

lynnvfrank committed:
Fixes network by whiching all the time

Signed-off-by: Franklin Webber <franklin.webber@gmail.com>

aaronlippold commented:
This is great progress. Watching with baited breath.

lynnvfrank committed:
Fixes the file_in_dir and re-use executable?

To replace Dir[] I built file_in_dir but I don't at this
point want to provide the same [] interface. So I changed
it to a method invocation.

Changed the the file_open back to return the StringIO object
I think that there was some other issue that was leading me
away from it when inside a block. Also I forgot that when you
examine each line of the StringIO you need to reset the cursor
that is keeping track of the position.

Decided to reuse the file_executable? in the which helper
which improves both.

Made notes about lacking windows support

Signed-off-by: Franklin Webber <franklin.webber@gmail.com>

lynnvfrank committed:
Updates timezone to use a linux command

This is no longer invoking the ruby way of getting the timezone

Signed-off-by: Franklin Webber <franklin.webber@gmail.com>

lynnvfrank committed:
etc works correctly in grabbing remote information

Instead of the ruby library we are resorting to parsing the files
found on the sytem.

Signed-off-by: Franklin Webber <franklin.webber@gmail.com>

lynnvfrank committed:
Updates root_group to use remote files

Signed-off-by: Franklin Webber <franklin.webber@gmail.com>

burtlo commented:
Alright I've made some great progress against a CentOS target. I did some [sorting](https://codeshack.io/json-sorter/) of the JSON to make sure I was comparing with not just my eyeballs. I added more todos, made a number of new fixes and conversions of existing plugins.

lynnvfrank committed:
Replaces ec2 with curl how slow!

Signed-off-by: Franklin Webber <franklin.webber@gmail.com>

lynnvfrank committed:
ohai plugins set the body to json?

There was no function and this is what you get with method_missing
Signed-off-by: Franklin Webber <franklin.webber@gmail.com>

lynnvfrank committed:
fixes which and virtualization

Signed-off-by: Franklin Webber <franklin.webber@gmail.com>

burtlo commented:
Alright I quickly worked on an HTTPClient wrapper for the `curl` on the system that makes `ec2`  and `cloud` possible #0d309ea. Found out why there was a `json?` attribute #731fb61. Fixed an issue with the `which` command to have a bit of better performance in some cases. Talked about an interface change in #cecf784.

lynnvfrank committed:
Updates chef_packages chef/ohai gathering

Signed-off-by: Franklin Webber <franklin.webber@gmail.com>

burtlo commented:
Here are all the commands that are being executed. I am including them here to provide some data for analysis later when returning for optimization work:

```
(uname -s)
(uname -m)
(test -f /etc/debian_version && cat /etc/debian_version)
(test -f /etc/os-release && cat /etc/os-release)
(test -f /usr/bin/FastCli)
(show version | json)
(test -f /etc/lsb-release && cat /etc/lsb-release)
(test -f /usr/bin/lsb-release && cat /usr/bin/lsb-release)
(test -f /etc/oracle-release && cat /etc/oracle-release)
(test -f /etc/enterprise-release && cat /etc/enterprise-release)
(test -f /etc/parallels-release && cat /etc/parallels-release)
(test -f /etc/system-release && cat /etc/system-release)
(test -f /etc/redhat-release && cat /etc/redhat-release)
(cat /proc/uptime || echo -n)
(which chef-client)
(/usr/bin/chef-client --version)
(find /opt -path '*/chef-14.11.21/lib')
(which ohai)
(/usr/bin/ohai --version)
(find /opt -path '*/ohai-14.8.10/lib')
(scala -version)
(groovy -v)
(rustc --version)
(php -v)
(ghc --version)
(ghci --version)
(cabal --version)
(stack --version)
(python -c "import sys; print (sys.version)")
(go version)
(gcc -v)
(/lib/libc.so.6)
(/lib64/libc.so.6)
(cc -V -flags)
(elixir -v)
(ruby -e "require 'rbconfig'; puts %Q(platform=#{RUBY_PLATFORM},version=#{RUBY_VERSION},release_date=#{RUBY_RELEASE_DATE},target=#{RbConfig::CONFIG['target']},target_cpu=#{RbConfig::CONFIG['target_cpu']},target_vendor=#{RbConfig::CONFIG['target_vendor']},target_os=#{RbConfig::CONFIG['target_os']},host=#{RbConfig::CONFIG['host']},host_cpu=#{RbConfig::CONFIG['host_cpu']},host_os=#{RbConfig::CONFIG['host_os']},host_vendor=#{RbConfig::CONFIG['host_vendor']},bin_dir=#{RbConfig::CONFIG['bindir']},ruby_bin=#{::File.join(RbConfig::CONFIG['bindir'], RbConfig::CONFIG['ruby_install_name'])},)")
(ruby -e "require 'rbconfig'; require 'rubygems'; puts ::Gem.default_exec_format % 'gem'")
(perl -V:version -V:archname)
(node -v)
(erl -eval '{ok, Ver} = file:read_file(filename:join([code:root_dir(), "releases", erlang:system_info(otp_release), "OTP_VERSION"])), Vsn = binary:bin_to_list(Ver, {0, byte_size(Ver) - 1}), io:format("~s,~s,~s", [Vsn, erlang:system_info(version), erlang:system_info(nif_version)]), halt().' -noshell)
(erl +V)
(which java)
(echo $PATH)
(stat -L /usr/local/bin/java 2>/dev/null --printf '%s
%f
%U
%u
%G
%g
%X
%Y
%C')
(stat -L /bin/java 2>/dev/null --printf '%s
%f
%U
%u
%G
%g
%X
%Y
%C')
(stat -L /usr/bin'
'/java 2>/dev/null --printf '%s
%f
%U
%u
%G
%g
%X
%Y
%C')
(stat -L /usr/bin/java 2>/dev/null --printf '%s
%f
%U
%u
%G
%g
%X
%Y
%C')
(stat -L /sbin/java 2>/dev/null --printf '%s
%f
%U
%u
%G
%g
%X
%Y
%C')
(stat -L /usr/sbin/java 2>/dev/null --printf '%s
%f
%U
%u
%G
%g
%X
%Y
%C')
[2019-04-03T23:01:19-05:00] WARN: Plugin Java: did not find java
(java -mx64m -version)
(lua -v)
(mono -V)
(which ip)
(echo $PATH)
(stat -L /usr/local/bin/ip 2>/dev/null --printf '%s
%f
%U
%u
%G
%g
%X
%Y
%C')
(stat -L /bin/ip 2>/dev/null --printf '%s
%f
%U
%u
%G
%g
%X
%Y
%C')
(stat -L /usr/bin'
'/ip 2>/dev/null --printf '%s
%f
%U
%u
%G
%g
%X
%Y
%C')
(stat -L /usr/bin/ip 2>/dev/null --printf '%s
%f
%U
%u
%G
%g
%X
%Y
%C')
(stat -L /sbin/ip 2>/dev/null --printf '%s
%f
%U
%u
%G
%g
%X
%Y
%C')
(test -e /proc/net/if_inet6)
(which ip)
(echo $PATH)
(/sbin/ip addr)
(which ip)
(echo $PATH)
(/sbin/ip -d -s link)
(which ip)
(echo $PATH)
(/sbin/ip -f inet neigh show)
(which ip)
(echo $PATH)
(/sbin/ip -o -f inet route show table main)
(test -e /proc/vz)
(test -e /proc/vz)
(which ip)
(echo $PATH)
(/sbin/ip -f inet6 neigh show)
(which ip)
(echo $PATH)
(/sbin/ip -o -f inet6 route show table main)
(test -e /proc/vz)
(test -e /proc/vz)
(test -e /proc/vz)
(test -e /proc/vz)
(test -e /proc/vz)
(test -e /proc/vz)
(test -e /proc/vz)
(test -e /proc/vz)
(test -e /proc/vz)
(test -e /proc/vz)
(which ethtool)
(echo $PATH)
(stat -L /usr/local/bin/ethtool 2>/dev/null --printf '%s
%f
%U
%u
%G
%g
%X
%Y
%C')
(stat -L /bin/ethtool 2>/dev/null --printf '%s
%f
%U
%u
%G
%g
%X
%Y
%C')
(stat -L /usr/bin'
'/ethtool 2>/dev/null --printf '%s
%f
%U
%u
%G
%g
%X
%Y
%C')
(stat -L /usr/bin/ethtool 2>/dev/null --printf '%s
%f
%U
%u
%G
%g
%X
%Y
%C')
(stat -L /sbin/ethtool 2>/dev/null --printf '%s
%f
%U
%u
%G
%g
%X
%Y
%C')
(/sbin/ethtool eth0)
(/sbin/ethtool -g eth0)
(uname -s)
(uname -r)
(uname -v)
(uname -m)
(uname -p)
(uname -o)
(env lsmod)
(which lsmod)
(echo $PATH)
(stat -L /usr/local/bin/lsmod 2>/dev/null --printf '%s
%f
%U
%u
%G
%g
%X
%Y
%C')
(stat -L /bin/lsmod 2>/dev/null --printf '%s
%f
%U
%u
%G
%g
%X
%Y
%C')
(stat -L /usr/bin'
'/lsmod 2>/dev/null --printf '%s
%f
%U
%u
%G
%g
%X
%Y
%C')
(stat -L /usr/bin/lsmod 2>/dev/null --printf '%s
%f
%U
%u
%G
%g
%X
%Y
%C')
(stat -L /sbin/lsmod 2>/dev/null --printf '%s
%f
%U
%u
%G
%g
%X
%Y
%C')
(/sbin/lsmod)
(test -e /sys/module/ipt_REJECT/version)
(test -e /sys/module/nf_conntrack_ipv4/version)
(test -e /sys/module/nf_defrag_ipv4/version)
(test -e /sys/module/iptable_filter/version)
(test -e /sys/module/ip_tables/version)
(test -e /sys/module/ip6t_REJECT/version)
(test -e /sys/module/nf_conntrack_ipv6/version)
(test -e /sys/module/nf_defrag_ipv6/version)
(test -e /sys/module/xt_state/version)
(test -e /sys/module/nf_conntrack/version)
(test -e /sys/module/ip6table_filter/version)
(test -e /sys/module/ip6_tables/version)
(test -e /sys/module/ib_ipoib/version)
(cat /sys/module/ib_ipoib/version || echo -n)
(test -e /sys/module/rdma_ucm/version)
(test -e /sys/module/ib_ucm/version)
(test -e /sys/module/ib_uverbs/version)
(test -e /sys/module/ib_umad/version)
(test -e /sys/module/rdma_cm/version)
(test -e /sys/module/ib_cm/version)
(test -e /sys/module/iw_cm/version)
(test -e /sys/module/ib_sa/version)
(test -e /sys/module/ib_mad/version)
(test -e /sys/module/ib_core/version)
(test -e /sys/module/ib_addr/version)
(test -e /sys/module/ipv6/version)
(test -e /sys/module/xen_netfront/version)
(test -e /sys/module/i2c_piix4/version)
(test -e /sys/module/i2c_core/version)
(test -e /sys/module/ext4/version)
(test -e /sys/module/jbd2/version)
(test -e /sys/module/mbcache/version)
(test -e /sys/module/xen_blkfront/version)
(test -e /sys/module/pata_acpi/version)
(cat /sys/module/pata_acpi/version || echo -n)
(test -e /sys/module/ata_generic/version)
(cat /sys/module/ata_generic/version || echo -n)
(test -e /sys/module/ata_piix/version)
(cat /sys/module/ata_piix/version || echo -n)
(test -e /sys/module/dm_mirror/version)
(test -e /sys/module/dm_region_hash/version)
(test -e /sys/module/dm_log/version)
(test -e /sys/module/dm_mod/version)
(test -e /usr/bin/lsb_release)
(test -e /etc/lsb-release)
(test -e /etc/os-release)
(test -e /etc/oracle-release)
(test -e /etc/enterprise-release)
(test -e /etc/f5-release)
(test -e /etc/debian_version)
(test -e /etc/parallels-release)
(test -e /etc/Eos-release)
(test -e /etc/redhat-release)
(cat /etc/redhat-release || echo -n)
(cat /proc/meminfo || echo -n)
(which dmidecode)
(echo $PATH)
(stat -L /usr/local/bin/dmidecode 2>/dev/null --printf '%s
%f
%U
%u
%G
%g
%X
%Y
%C')
(stat -L /bin/dmidecode 2>/dev/null --printf '%s
%f
%U
%u
%G
%g
%X
%Y
%C')
(stat -L /usr/bin'
'/dmidecode 2>/dev/null --printf '%s
%f
%U
%u
%G
%g
%X
%Y
%C')
(stat -L /usr/bin/dmidecode 2>/dev/null --printf '%s
%f
%U
%u
%G
%g
%X
%Y
%C')
(stat -L /sbin/dmidecode 2>/dev/null --printf '%s
%f
%U
%u
%G
%g
%X
%Y
%C')
(stat -L /usr/sbin/dmidecode 2>/dev/null --printf '%s
%f
%U
%u
%G
%g
%X
%Y
%C')
(/usr/sbin/dmidecode)
(hostname -s)
(hostname)
(hostname --fqdn)
(hostname --fqdn)
(test -e /proc/1/comm)
(cat /proc/1/comm || echo -n)
(VBoxControl guestproperty enumerate)
(date +%Z)
(xenstore-read vm-data/provider_data/provider)
(test -e /sys/hypervisor/uuid)
(cat /sys/hypervisor/uuid || echo -n)
(curl -Is 169.254.169.254:80 --connect-timeout 2)
(curl -i http://169.254.169.254:80/ --keepalive-time 10)
(curl -i http://169.254.169.254:80/2016-09-02/meta-data/ --keepalive-time 10)
(curl -i http://169.254.169.254:80/2016-09-02/meta-data/ami-id --keepalive-time 10)
(curl -i http://169.254.169.254:80/2016-09-02/meta-data/ami-launch-index --keepalive-time 10)
(curl -i http://169.254.169.254:80/2016-09-02/meta-data/ami-manifest-path --keepalive-time 10)
(curl -i http://169.254.169.254:80/2016-09-02/meta-data/block-device-mapping/ --keepalive-time 10)
(curl -i http://169.254.169.254:80/2016-09-02/meta-data/block-device-mapping/ami --keepalive-time 10)
(curl -i http://169.254.169.254:80/2016-09-02/meta-data/block-device-mapping/root --keepalive-time 10)
(curl -i http://169.254.169.254:80/2016-09-02/meta-data/hostname --keepalive-time 10)
(curl -i http://169.254.169.254:80/2016-09-02/meta-data/instance-action --keepalive-time 10)
(curl -i http://169.254.169.254:80/2016-09-02/meta-data/instance-id --keepalive-time 10)
(curl -i http://169.254.169.254:80/2016-09-02/meta-data/instance-type --keepalive-time 10)
(curl -i http://169.254.169.254:80/2016-09-02/meta-data/local-hostname --keepalive-time 10)
(curl -i http://169.254.169.254:80/2016-09-02/meta-data/local-ipv4 --keepalive-time 10)
(curl -i http://169.254.169.254:80/2016-09-02/meta-data/mac --keepalive-time 10)
(curl -i http://169.254.169.254:80/2016-09-02/meta-data/metrics/ --keepalive-time 10)
(curl -i http://169.254.169.254:80/2016-09-02/meta-data/metrics/vhostmd --keepalive-time 10)
(curl -i http://169.254.169.254:80/2016-09-02/meta-data/network/ --keepalive-time 10)
(curl -i http://169.254.169.254:80/2016-09-02/meta-data/network/interfaces/ --keepalive-time 10)
(curl -i http://169.254.169.254:80/2016-09-02/meta-data/network/interfaces/macs/ --keepalive-time 10)
(curl -i http://169.254.169.254:80/2016-09-02/meta-data/network/interfaces/macs/0a:f3:15:99:19:36/ --keepalive-time 10)
(curl -i http://169.254.169.254:80/2016-09-02/meta-data/network/interfaces/macs/0a:f3:15:99:19:36/device-number --keepalive-time 10)
(curl -i http://169.254.169.254:80/2016-09-02/meta-data/network/interfaces/macs/0a:f3:15:99:19:36/interface-id --keepalive-time 10)
(curl -i http://169.254.169.254:80/2016-09-02/meta-data/network/interfaces/macs/0a:f3:15:99:19:36/ipv4-associations/ --keepalive-time 10)
(curl -i http://169.254.169.254:80/2016-09-02/meta-data/network/interfaces/macs/0a:f3:15:99:19:36/ipv4-associations/54.167.86.159 --keepalive-time 10)
(curl -i http://169.254.169.254:80/2016-09-02/meta-data/network/interfaces/macs/0a:f3:15:99:19:36/local-hostname --keepalive-time 10)
(curl -i http://169.254.169.254:80/2016-09-02/meta-data/network/interfaces/macs/0a:f3:15:99:19:36/local-ipv4s --keepalive-time 10)
(curl -i http://169.254.169.254:80/2016-09-02/meta-data/network/interfaces/macs/0a:f3:15:99:19:36/mac --keepalive-time 10)
(curl -i http://169.254.169.254:80/2016-09-02/meta-data/network/interfaces/macs/0a:f3:15:99:19:36/owner-id --keepalive-time 10)
(curl -i http://169.254.169.254:80/2016-09-02/meta-data/network/interfaces/macs/0a:f3:15:99:19:36/public-hostname --keepalive-time 10)
(curl -i http://169.254.169.254:80/2016-09-02/meta-data/network/interfaces/macs/0a:f3:15:99:19:36/public-ipv4s --keepalive-time 10)
(curl -i http://169.254.169.254:80/2016-09-02/meta-data/network/interfaces/macs/0a:f3:15:99:19:36/security-group-ids --keepalive-time 10)
(curl -i http://169.254.169.254:80/2016-09-02/meta-data/network/interfaces/macs/0a:f3:15:99:19:36/security-groups --keepalive-time 10)
(curl -i http://169.254.169.254:80/2016-09-02/meta-data/network/interfaces/macs/0a:f3:15:99:19:36/subnet-id --keepalive-time 10)
(curl -i http://169.254.169.254:80/2016-09-02/meta-data/network/interfaces/macs/0a:f3:15:99:19:36/subnet-ipv4-cidr-block --keepalive-time 10)
(curl -i http://169.254.169.254:80/2016-09-02/meta-data/network/interfaces/macs/0a:f3:15:99:19:36/vpc-id --keepalive-time 10)
(curl -i http://169.254.169.254:80/2016-09-02/meta-data/network/interfaces/macs/0a:f3:15:99:19:36/vpc-ipv4-cidr-block --keepalive-time 10)
(curl -i http://169.254.169.254:80/2016-09-02/meta-data/network/interfaces/macs/0a:f3:15:99:19:36/vpc-ipv4-cidr-blocks --keepalive-time 10)
(curl -i http://169.254.169.254:80/2016-09-02/meta-data/placement/ --keepalive-time 10)
(curl -i http://169.254.169.254:80/2016-09-02/meta-data/placement/availability-zone --keepalive-time 10)
(curl -i http://169.254.169.254:80/2016-09-02/meta-data/product-codes --keepalive-time 10)
(curl -i http://169.254.169.254:80/2016-09-02/meta-data/profile --keepalive-time 10)
(curl -i http://169.254.169.254:80/2016-09-02/meta-data/public-hostname --keepalive-time 10)
(curl -i http://169.254.169.254:80/2016-09-02/meta-data/public-ipv4 --keepalive-time 10)
(curl -i http://169.254.169.254:80/2016-09-02/meta-data/public-keys/ --keepalive-time 10)
(curl -i http://169.254.169.254:80/2016-09-02/meta-data/public-keys/0/ --keepalive-time 10)
(curl -i http://169.254.169.254:80/2016-09-02/meta-data/public-keys/0/openssh-key --keepalive-time 10)
(curl -i http://169.254.169.254:80/2016-09-02/meta-data/reservation-id --keepalive-time 10)
(curl -i http://169.254.169.254:80/2016-09-02/meta-data/security-groups --keepalive-time 10)
(curl -i http://169.254.169.254:80/2016-09-02/meta-data/services/ --keepalive-time 10)
(curl -i http://169.254.169.254:80/2016-09-02/meta-data/services/domain --keepalive-time 10)
(curl -i http://169.254.169.254:80/2016-09-02/meta-data/services/partition --keepalive-time 10)
(curl -i http://169.254.169.254:80/2016-09-02/user-data/ --keepalive-time 10)
(curl -i http://169.254.169.254:80/2016-09-02/dynamic/instance-identity/document/ --keepalive-time 10)
(cat /etc/passwd || echo -n)
(cat /etc/group || echo -n)
(which docker)
(echo $PATH)
(stat -L /usr/local/bin/docker 2>/dev/null --printf '%s
%f
%U
%u
%G
%g
%X
%Y
%C')
(stat -L /bin/docker 2>/dev/null --printf '%s
%f
%U
%u
%G
%g
%X
%Y
%C')
(stat -L /usr/bin'
'/docker 2>/dev/null --printf '%s
%f
%U
%u
%G
%g
%X
%Y
%C')
(stat -L /usr/bin/docker 2>/dev/null --printf '%s
%f
%U
%u
%G
%g
%X
%Y
%C')
(stat -L /sbin/docker 2>/dev/null --printf '%s
%f
%U
%u
%G
%g
%X
%Y
%C')
(stat -L /usr/sbin/docker 2>/dev/null --printf '%s
%f
%U
%u
%G
%g
%X
%Y
%C')
[2019-04-03T23:01:56-05:00] WARN: Plugin Virtualization: did not find docker
(test -e /proc/xen)
(test -e /proc/xen/capabilities)
(test -e /proc/modules)
(cat /proc/modules || echo -n)
(which nova)
(echo $PATH)
(stat -L /usr/local/bin/nova 2>/dev/null --printf '%s
%f
%U
%u
%G
%g
%X
%Y
%C')
(stat -L /bin/nova 2>/dev/null --printf '%s
%f
%U
%u
%G
%g
%X
%Y
%C')
(stat -L /usr/bin'
'/nova 2>/dev/null --printf '%s
%f
%U
%u
%G
%g
%X
%Y
%C')
(stat -L /usr/bin/nova 2>/dev/null --printf '%s
%f
%U
%u
%G
%g
%X
%Y
%C')
(stat -L /sbin/nova 2>/dev/null --printf '%s
%f
%U
%u
%G
%g
%X
%Y
%C')
(stat -L /usr/sbin/nova 2>/dev/null --printf '%s
%f
%U
%u
%G
%g
%X
%Y
%C')
[2019-04-03T23:01:58-05:00] WARN: Plugin Virtualization: did not find nova
(test -e /proc/cpuinfo)
(cat /proc/cpuinfo || echo -n)
(test -e /sys/devices/virtual/misc/kvm)
(test -e /proc/bc/0)
(test -e /proc/vz)
(test -e /var/lib/hyperv/.kvp_pool_3)
(test -e /proc/self/status)
(cat /proc/self/status || echo -n)
(test -e /proc/self/cgroup)
(cat /proc/self/cgroup || echo -n)
(stat -L /proc/self/cgroup 2>/dev/null --printf '%s
%f
%U
%u
%G
%g
%X
%Y
%C')
(cat /proc/1/environ || echo -n)
(stat -L /proc/1/environ 2>/dev/null --printf '%s
%f
%U
%u
%G
%g
%X
%Y
%C')
(which lxc-version)
(echo $PATH)
(stat -L /usr/local/bin/lxc-version 2>/dev/null --printf '%s
%f
%U
%u
%G
%g
%X
%Y
%C')
(stat -L /bin/lxc-version 2>/dev/null --printf '%s
%f
%U
%u
%G
%g
%X
%Y
%C')
(stat -L /usr/bin'
'/lxc-version 2>/dev/null --printf '%s
%f
%U
%u
%G
%g
%X
%Y
%C')
(stat -L /usr/bin/lxc-version 2>/dev/null --printf '%s
%f
%U
%u
%G
%g
%X
%Y
%C')
(stat -L /sbin/lxc-version 2>/dev/null --printf '%s
%f
%U
%u
%G
%g
%X
%Y
%C')
(stat -L /usr/sbin/lxc-version 2>/dev/null --printf '%s
%f
%U
%u
%G
%g
%X
%Y
%C')
[2019-04-03T23:02:01-05:00] WARN: Plugin Virtualization: did not find lxc-version
(which lxc-start)
(echo $PATH)
(stat -L /usr/local/bin/lxc-start 2>/dev/null --printf '%s
%f
%U
%u
%G
%g
%X
%Y
%C')
(stat -L /bin/lxc-start 2>/dev/null --printf '%s
%f
%U
%u
%G
%g
%X
%Y
%C')
(stat -L /usr/bin'
'/lxc-start 2>/dev/null --printf '%s
%f
%U
%u
%G
%g
%X
%Y
%C')
(stat -L /usr/bin/lxc-start 2>/dev/null --printf '%s
%f
%U
%u
%G
%g
%X
%Y
%C')
(stat -L /sbin/lxc-start 2>/dev/null --printf '%s
%f
%U
%u
%G
%g
%X
%Y
%C')
(stat -L /usr/sbin/lxc-start 2>/dev/null --printf '%s
%f
%U
%u
%G
%g
%X
%Y
%C')
[2019-04-03T23:02:03-05:00] WARN: Plugin Virtualization: did not find lxc-start
(test -e /dev/lxd/sock)
(test -e /var/lib/lxd/devlxd)
(test -e /var/snap/lxd/common/lxd/devlxd)
(test -e /etc/machine-id)
(test -e /var/lib/dbus/machine-id)
(test -e /etc/ssh/sshd_config)
(cat /etc/ssh/sshd_config || echo -n)
(test -e /etc/ssh/ssh_host_dsa_key.pub)
(cat /etc/ssh/ssh_host_dsa_key.pub || echo -n)
(test -e /etc/ssh/ssh_host_rsa_key.pub)
(cat /etc/ssh/ssh_host_rsa_key.pub || echo -n)
(test -e /etc/ssh/ssh_host_ecdsa_key.pub)
(test -e /etc/ssh/ssh_host_ed25519_key.pub)
(rpm -qa --qf '%{NAME}\t%|EPOCH?{%{EPOCH}}:{0}|\t%{VERSION}\t%{RELEASE}\t%{INSTALLTIME}\t%{ARCH}\n')
(test -e /etc/shells)
(cat /etc/shells || echo -n)
(test -e /sys/class/dmi/id/product_name)
(cat /sys/class/dmi/id/product_name || echo -n)
(whoami)
(docker info --format '{{json .}}')
(test -e /proc/mdstat)
(cat /proc/mdstat || echo -n)
(test -e /sys/block)
(ls -d /sys/block/*)
(test -e /sys/block/loop0/size)
(cat /sys/block/loop0/size || echo -n)
(test -e /sys/block/loop0/removable)
(cat /sys/block/loop0/removable || echo -n)
(test -e /sys/block/loop0/device/model)
(test -e /sys/block/loop0/device/rev)
(test -e /sys/block/loop0/device/state)
(test -e /sys/block/loop0/device/timeout)
(test -e /sys/block/loop0/device/vendor)
(test -e /sys/block/loop0/device/queue_depth)
(test -e /sys/block/loop0/queue/rotational)
(cat /sys/block/loop0/queue/rotational || echo -n)
(test -e /sys/block/loop0/queue/physical_block_size)
(cat /sys/block/loop0/queue/physical_block_size || echo -n)
(test -e /sys/block/loop0/queue/logical_block_size)
(cat /sys/block/loop0/queue/logical_block_size || echo -n)
(test -e /sys/block/loop1/size)
(cat /sys/block/loop1/size || echo -n)
(test -e /sys/block/loop1/removable)
(cat /sys/block/loop1/removable || echo -n)
(test -e /sys/block/loop1/device/model)
(test -e /sys/block/loop1/device/rev)
(test -e /sys/block/loop1/device/state)
(test -e /sys/block/loop1/device/timeout)
(test -e /sys/block/loop1/device/vendor)
(test -e /sys/block/loop1/device/queue_depth)
(test -e /sys/block/loop1/queue/rotational)
(cat /sys/block/loop1/queue/rotational || echo -n)
(test -e /sys/block/loop1/queue/physical_block_size)
(cat /sys/block/loop1/queue/physical_block_size || echo -n)
(test -e /sys/block/loop1/queue/logical_block_size)
(cat /sys/block/loop1/queue/logical_block_size || echo -n)
(test -e /sys/block/loop2/size)
(cat /sys/block/loop2/size || echo -n)
(test -e /sys/block/loop2/removable)
(cat /sys/block/loop2/removable || echo -n)
(test -e /sys/block/loop2/device/model)
(test -e /sys/block/loop2/device/rev)
(test -e /sys/block/loop2/device/state)
(test -e /sys/block/loop2/device/timeout)
(test -e /sys/block/loop2/device/vendor)
(test -e /sys/block/loop2/device/queue_depth)
(test -e /sys/block/loop2/queue/rotational)
(cat /sys/block/loop2/queue/rotational || echo -n)
(test -e /sys/block/loop2/queue/physical_block_size)
(cat /sys/block/loop2/queue/physical_block_size || echo -n)
(test -e /sys/block/loop2/queue/logical_block_size)
(cat /sys/block/loop2/queue/logical_block_size || echo -n)
(test -e /sys/block/loop3/size)
(cat /sys/block/loop3/size || echo -n)
(test -e /sys/block/loop3/removable)
(cat /sys/block/loop3/removable || echo -n)
(test -e /sys/block/loop3/device/model)
(test -e /sys/block/loop3/device/rev)
(test -e /sys/block/loop3/device/state)
(test -e /sys/block/loop3/device/timeout)
(test -e /sys/block/loop3/device/vendor)
(test -e /sys/block/loop3/device/queue_depth)
(test -e /sys/block/loop3/queue/rotational)
(cat /sys/block/loop3/queue/rotational || echo -n)
(test -e /sys/block/loop3/queue/physical_block_size)
(cat /sys/block/loop3/queue/physical_block_size || echo -n)
(test -e /sys/block/loop3/queue/logical_block_size)
(cat /sys/block/loop3/queue/logical_block_size || echo -n)
(test -e /sys/block/loop4/size)
(cat /sys/block/loop4/size || echo -n)
(test -e /sys/block/loop4/removable)
(cat /sys/block/loop4/removable || echo -n)
(test -e /sys/block/loop4/device/model)
(test -e /sys/block/loop4/device/rev)
(test -e /sys/block/loop4/device/state)
(test -e /sys/block/loop4/device/timeout)
(test -e /sys/block/loop4/device/vendor)
(test -e /sys/block/loop4/device/queue_depth)
(test -e /sys/block/loop4/queue/rotational)
(cat /sys/block/loop4/queue/rotational || echo -n)
(test -e /sys/block/loop4/queue/physical_block_size)
(cat /sys/block/loop4/queue/physical_block_size || echo -n)
(test -e /sys/block/loop4/queue/logical_block_size)
(cat /sys/block/loop4/queue/logical_block_size || echo -n)
(test -e /sys/block/loop5/size)
(cat /sys/block/loop5/size || echo -n)
(test -e /sys/block/loop5/removable)
(cat /sys/block/loop5/removable || echo -n)
(test -e /sys/block/loop5/device/model)
(test -e /sys/block/loop5/device/rev)
(test -e /sys/block/loop5/device/state)
(test -e /sys/block/loop5/device/timeout)
(test -e /sys/block/loop5/device/vendor)
(test -e /sys/block/loop5/device/queue_depth)
(test -e /sys/block/loop5/queue/rotational)
(cat /sys/block/loop5/queue/rotational || echo -n)
(test -e /sys/block/loop5/queue/physical_block_size)
(cat /sys/block/loop5/queue/physical_block_size || echo -n)
(test -e /sys/block/loop5/queue/logical_block_size)
(cat /sys/block/loop5/queue/logical_block_size || echo -n)
(test -e /sys/block/loop6/size)
(cat /sys/block/loop6/size || echo -n)
(test -e /sys/block/loop6/removable)
(cat /sys/block/loop6/removable || echo -n)
(test -e /sys/block/loop6/device/model)
(test -e /sys/block/loop6/device/rev)
(test -e /sys/block/loop6/device/state)
(test -e /sys/block/loop6/device/timeout)
(test -e /sys/block/loop6/device/vendor)
(test -e /sys/block/loop6/device/queue_depth)
(test -e /sys/block/loop6/queue/rotational)
(cat /sys/block/loop6/queue/rotational || echo -n)
(test -e /sys/block/loop6/queue/physical_block_size)
(cat /sys/block/loop6/queue/physical_block_size || echo -n)
(test -e /sys/block/loop6/queue/logical_block_size)
(cat /sys/block/loop6/queue/logical_block_size || echo -n)
(test -e /sys/block/loop7/size)
(cat /sys/block/loop7/size || echo -n)
(test -e /sys/block/loop7/removable)
(cat /sys/block/loop7/removable || echo -n)
(test -e /sys/block/loop7/device/model)
(test -e /sys/block/loop7/device/rev)
(test -e /sys/block/loop7/device/state)
(test -e /sys/block/loop7/device/timeout)
(test -e /sys/block/loop7/device/vendor)
(test -e /sys/block/loop7/device/queue_depth)
(test -e /sys/block/loop7/queue/rotational)
(cat /sys/block/loop7/queue/rotational || echo -n)
(test -e /sys/block/loop7/queue/physical_block_size)
(cat /sys/block/loop7/queue/physical_block_size || echo -n)
(test -e /sys/block/loop7/queue/logical_block_size)
(cat /sys/block/loop7/queue/logical_block_size || echo -n)
(test -e /sys/block/ram0/size)
(cat /sys/block/ram0/size || echo -n)
(test -e /sys/block/ram0/removable)
(cat /sys/block/ram0/removable || echo -n)
(test -e /sys/block/ram0/device/model)
(test -e /sys/block/ram0/device/rev)
(test -e /sys/block/ram0/device/state)
(test -e /sys/block/ram0/device/timeout)
(test -e /sys/block/ram0/device/vendor)
(test -e /sys/block/ram0/device/queue_depth)
(test -e /sys/block/ram0/queue/rotational)
(cat /sys/block/ram0/queue/rotational || echo -n)
(test -e /sys/block/ram0/queue/physical_block_size)
(cat /sys/block/ram0/queue/physical_block_size || echo -n)
(test -e /sys/block/ram0/queue/logical_block_size)
(cat /sys/block/ram0/queue/logical_block_size || echo -n)
(test -e /sys/block/ram1/size)
(cat /sys/block/ram1/size || echo -n)
(test -e /sys/block/ram1/removable)
(cat /sys/block/ram1/removable || echo -n)
(test -e /sys/block/ram1/device/model)
(test -e /sys/block/ram1/device/rev)
(test -e /sys/block/ram1/device/state)
(test -e /sys/block/ram1/device/timeout)
(test -e /sys/block/ram1/device/vendor)
(test -e /sys/block/ram1/device/queue_depth)
(test -e /sys/block/ram1/queue/rotational)
(cat /sys/block/ram1/queue/rotational || echo -n)
(test -e /sys/block/ram1/queue/physical_block_size)
(cat /sys/block/ram1/queue/physical_block_size || echo -n)
(test -e /sys/block/ram1/queue/logical_block_size)
(cat /sys/block/ram1/queue/logical_block_size || echo -n)
(test -e /sys/block/ram10/size)
(cat /sys/block/ram10/size || echo -n)
(test -e /sys/block/ram10/removable)
(cat /sys/block/ram10/removable || echo -n)
(test -e /sys/block/ram10/device/model)
(test -e /sys/block/ram10/device/rev)
(test -e /sys/block/ram10/device/state)
(test -e /sys/block/ram10/device/timeout)
(test -e /sys/block/ram10/device/vendor)
(test -e /sys/block/ram10/device/queue_depth)
(test -e /sys/block/ram10/queue/rotational)
(cat /sys/block/ram10/queue/rotational || echo -n)
(test -e /sys/block/ram10/queue/physical_block_size)
(cat /sys/block/ram10/queue/physical_block_size || echo -n)
(test -e /sys/block/ram10/queue/logical_block_size)
(cat /sys/block/ram10/queue/logical_block_size || echo -n)
(test -e /sys/block/ram11/size)
(cat /sys/block/ram11/size || echo -n)
(test -e /sys/block/ram11/removable)
(cat /sys/block/ram11/removable || echo -n)
(test -e /sys/block/ram11/device/model)
(test -e /sys/block/ram11/device/rev)
(test -e /sys/block/ram11/device/state)
(test -e /sys/block/ram11/device/timeout)
(test -e /sys/block/ram11/device/vendor)
(test -e /sys/block/ram11/device/queue_depth)
(test -e /sys/block/ram11/queue/rotational)
(cat /sys/block/ram11/queue/rotational || echo -n)
(test -e /sys/block/ram11/queue/physical_block_size)
(cat /sys/block/ram11/queue/physical_block_size || echo -n)
(test -e /sys/block/ram11/queue/logical_block_size)
(cat /sys/block/ram11/queue/logical_block_size || echo -n)
(test -e /sys/block/ram12/size)
(cat /sys/block/ram12/size || echo -n)
(test -e /sys/block/ram12/removable)
(cat /sys/block/ram12/removable || echo -n)
(test -e /sys/block/ram12/device/model)
(test -e /sys/block/ram12/device/rev)
(test -e /sys/block/ram12/device/state)
(test -e /sys/block/ram12/device/timeout)
(test -e /sys/block/ram12/device/vendor)
(test -e /sys/block/ram12/device/queue_depth)
(test -e /sys/block/ram12/queue/rotational)
(cat /sys/block/ram12/queue/rotational || echo -n)
(test -e /sys/block/ram12/queue/physical_block_size)
(cat /sys/block/ram12/queue/physical_block_size || echo -n)
(test -e /sys/block/ram12/queue/logical_block_size)
(cat /sys/block/ram12/queue/logical_block_size || echo -n)
(test -e /sys/block/ram13/size)
(cat /sys/block/ram13/size || echo -n)
(test -e /sys/block/ram13/removable)
(cat /sys/block/ram13/removable || echo -n)
(test -e /sys/block/ram13/device/model)
(test -e /sys/block/ram13/device/rev)
(test -e /sys/block/ram13/device/state)
(test -e /sys/block/ram13/device/timeout)
(test -e /sys/block/ram13/device/vendor)
(test -e /sys/block/ram13/device/queue_depth)
(test -e /sys/block/ram13/queue/rotational)
(cat /sys/block/ram13/queue/rotational || echo -n)
(test -e /sys/block/ram13/queue/physical_block_size)
(cat /sys/block/ram13/queue/physical_block_size || echo -n)
(test -e /sys/block/ram13/queue/logical_block_size)
(cat /sys/block/ram13/queue/logical_block_size || echo -n)
(test -e /sys/block/ram14/size)
(cat /sys/block/ram14/size || echo -n)
(test -e /sys/block/ram14/removable)
(cat /sys/block/ram14/removable || echo -n)
(test -e /sys/block/ram14/device/model)
(test -e /sys/block/ram14/device/rev)
(test -e /sys/block/ram14/device/state)
(test -e /sys/block/ram14/device/timeout)
(test -e /sys/block/ram14/device/vendor)
(test -e /sys/block/ram14/device/queue_depth)
(test -e /sys/block/ram14/queue/rotational)
(cat /sys/block/ram14/queue/rotational || echo -n)
(test -e /sys/block/ram14/queue/physical_block_size)
(cat /sys/block/ram14/queue/physical_block_size || echo -n)
(test -e /sys/block/ram14/queue/logical_block_size)
(cat /sys/block/ram14/queue/logical_block_size || echo -n)
(test -e /sys/block/ram15/size)
(cat /sys/block/ram15/size || echo -n)
(test -e /sys/block/ram15/removable)
(cat /sys/block/ram15/removable || echo -n)
(test -e /sys/block/ram15/device/model)
(test -e /sys/block/ram15/device/rev)
(test -e /sys/block/ram15/device/state)
(test -e /sys/block/ram15/device/timeout)
(test -e /sys/block/ram15/device/vendor)
(test -e /sys/block/ram15/device/queue_depth)
(test -e /sys/block/ram15/queue/rotational)
(cat /sys/block/ram15/queue/rotational || echo -n)
(test -e /sys/block/ram15/queue/physical_block_size)
(cat /sys/block/ram15/queue/physical_block_size || echo -n)
(test -e /sys/block/ram15/queue/logical_block_size)
(cat /sys/block/ram15/queue/logical_block_size || echo -n)
(test -e /sys/block/ram2/size)
(cat /sys/block/ram2/size || echo -n)
(test -e /sys/block/ram2/removable)
(cat /sys/block/ram2/removable || echo -n)
(test -e /sys/block/ram2/device/model)
(test -e /sys/block/ram2/device/rev)
(test -e /sys/block/ram2/device/state)
(test -e /sys/block/ram2/device/timeout)
(test -e /sys/block/ram2/device/vendor)
(test -e /sys/block/ram2/device/queue_depth)
(test -e /sys/block/ram2/queue/rotational)
(cat /sys/block/ram2/queue/rotational || echo -n)
(test -e /sys/block/ram2/queue/physical_block_size)
(cat /sys/block/ram2/queue/physical_block_size || echo -n)
(test -e /sys/block/ram2/queue/logical_block_size)
(cat /sys/block/ram2/queue/logical_block_size || echo -n)
(test -e /sys/block/ram3/size)
(cat /sys/block/ram3/size || echo -n)
(test -e /sys/block/ram3/removable)
(cat /sys/block/ram3/removable || echo -n)
(test -e /sys/block/ram3/device/model)
(test -e /sys/block/ram3/device/rev)
(test -e /sys/block/ram3/device/state)
(test -e /sys/block/ram3/device/timeout)
(test -e /sys/block/ram3/device/vendor)
(test -e /sys/block/ram3/device/queue_depth)
(test -e /sys/block/ram3/queue/rotational)
(cat /sys/block/ram3/queue/rotational || echo -n)
(test -e /sys/block/ram3/queue/physical_block_size)
(cat /sys/block/ram3/queue/physical_block_size || echo -n)
(test -e /sys/block/ram3/queue/logical_block_size)
(cat /sys/block/ram3/queue/logical_block_size || echo -n)
(test -e /sys/block/ram4/size)
(cat /sys/block/ram4/size || echo -n)
(test -e /sys/block/ram4/removable)
(cat /sys/block/ram4/removable || echo -n)
(test -e /sys/block/ram4/device/model)
(test -e /sys/block/ram4/device/rev)
(test -e /sys/block/ram4/device/state)
(test -e /sys/block/ram4/device/timeout)
(test -e /sys/block/ram4/device/vendor)
(test -e /sys/block/ram4/device/queue_depth)
(test -e /sys/block/ram4/queue/rotational)
(cat /sys/block/ram4/queue/rotational || echo -n)
(test -e /sys/block/ram4/queue/physical_block_size)
(cat /sys/block/ram4/queue/physical_block_size || echo -n)
(test -e /sys/block/ram4/queue/logical_block_size)
(cat /sys/block/ram4/queue/logical_block_size || echo -n)
(test -e /sys/block/ram5/size)
(cat /sys/block/ram5/size || echo -n)
(test -e /sys/block/ram5/removable)
(cat /sys/block/ram5/removable || echo -n)
(test -e /sys/block/ram5/device/model)
(test -e /sys/block/ram5/device/rev)
(test -e /sys/block/ram5/device/state)
(test -e /sys/block/ram5/device/timeout)
(test -e /sys/block/ram5/device/vendor)
(test -e /sys/block/ram5/device/queue_depth)
(test -e /sys/block/ram5/queue/rotational)
(cat /sys/block/ram5/queue/rotational || echo -n)
(test -e /sys/block/ram5/queue/physical_block_size)
(cat /sys/block/ram5/queue/physical_block_size || echo -n)
(test -e /sys/block/ram5/queue/logical_block_size)
(cat /sys/block/ram5/queue/logical_block_size || echo -n)
(test -e /sys/block/ram6/size)
(cat /sys/block/ram6/size || echo -n)
(test -e /sys/block/ram6/removable)
(cat /sys/block/ram6/removable || echo -n)
(test -e /sys/block/ram6/device/model)
(test -e /sys/block/ram6/device/rev)
(test -e /sys/block/ram6/device/state)
(test -e /sys/block/ram6/device/timeout)
(test -e /sys/block/ram6/device/vendor)
(test -e /sys/block/ram6/device/queue_depth)
(test -e /sys/block/ram6/queue/rotational)
(cat /sys/block/ram6/queue/rotational || echo -n)
(test -e /sys/block/ram6/queue/physical_block_size)
(cat /sys/block/ram6/queue/physical_block_size || echo -n)
(test -e /sys/block/ram6/queue/logical_block_size)
(cat /sys/block/ram6/queue/logical_block_size || echo -n)
(test -e /sys/block/ram7/size)
(cat /sys/block/ram7/size || echo -n)
(test -e /sys/block/ram7/removable)
(cat /sys/block/ram7/removable || echo -n)
(test -e /sys/block/ram7/device/model)
(test -e /sys/block/ram7/device/rev)
(test -e /sys/block/ram7/device/state)
(test -e /sys/block/ram7/device/timeout)
(test -e /sys/block/ram7/device/vendor)
(test -e /sys/block/ram7/device/queue_depth)
(test -e /sys/block/ram7/queue/rotational)
(cat /sys/block/ram7/queue/rotational || echo -n)
(test -e /sys/block/ram7/queue/physical_block_size)
(cat /sys/block/ram7/queue/physical_block_size || echo -n)
(test -e /sys/block/ram7/queue/logical_block_size)
(cat /sys/block/ram7/queue/logical_block_size || echo -n)
(test -e /sys/block/ram8/size)
(cat /sys/block/ram8/size || echo -n)
(test -e /sys/block/ram8/removable)
(cat /sys/block/ram8/removable || echo -n)
(test -e /sys/block/ram8/device/model)
(test -e /sys/block/ram8/device/rev)
(test -e /sys/block/ram8/device/state)
(test -e /sys/block/ram8/device/timeout)
(test -e /sys/block/ram8/device/vendor)
(test -e /sys/block/ram8/device/queue_depth)
(test -e /sys/block/ram8/queue/rotational)
(cat /sys/block/ram8/queue/rotational || echo -n)
(test -e /sys/block/ram8/queue/physical_block_size)
(cat /sys/block/ram8/queue/physical_block_size || echo -n)
(test -e /sys/block/ram8/queue/logical_block_size)
(cat /sys/block/ram8/queue/logical_block_size || echo -n)
(test -e /sys/block/ram9/size)
(cat /sys/block/ram9/size || echo -n)
(test -e /sys/block/ram9/removable)
(cat /sys/block/ram9/removable || echo -n)
(test -e /sys/block/ram9/device/model)
(test -e /sys/block/ram9/device/rev)
(test -e /sys/block/ram9/device/state)
(test -e /sys/block/ram9/device/timeout)
(test -e /sys/block/ram9/device/vendor)
(test -e /sys/block/ram9/device/queue_depth)
(test -e /sys/block/ram9/queue/rotational)
(cat /sys/block/ram9/queue/rotational || echo -n)
(test -e /sys/block/ram9/queue/physical_block_size)
(cat /sys/block/ram9/queue/physical_block_size || echo -n)
(test -e /sys/block/ram9/queue/logical_block_size)
(cat /sys/block/ram9/queue/logical_block_size || echo -n)
(test -e /sys/block/xvda/size)
(cat /sys/block/xvda/size || echo -n)
(test -e /sys/block/xvda/removable)
(cat /sys/block/xvda/removable || echo -n)
(test -e /sys/block/xvda/device/model)
(test -e /sys/block/xvda/device/rev)
(test -e /sys/block/xvda/device/state)
(test -e /sys/block/xvda/device/timeout)
(test -e /sys/block/xvda/device/vendor)
(test -e /sys/block/xvda/device/queue_depth)
(test -e /sys/block/xvda/queue/rotational)
(cat /sys/block/xvda/queue/rotational || echo -n)
(test -e /sys/block/xvda/queue/physical_block_size)
(cat /sys/block/xvda/queue/physical_block_size || echo -n)
(test -e /sys/block/xvda/queue/logical_block_size)
(cat /sys/block/xvda/queue/logical_block_size || echo -n)
(which hostnamectl)
(echo $PATH)
(stat -L /usr/local/bin/hostnamectl 2>/dev/null --printf '%s
%f
%U
%u
%G
%g
%X
%Y
%C')
(stat -L /bin/hostnamectl 2>/dev/null --printf '%s
%f
%U
%u
%G
%g
%X
%Y
%C')
(stat -L /usr/bin'
'/hostnamectl 2>/dev/null --printf '%s
%f
%U
%u
%G
%g
%X
%Y
%C')
(stat -L /usr/bin/hostnamectl 2>/dev/null --printf '%s
%f
%U
%u
%G
%g
%X
%Y
%C')
(stat -L /sbin/hostnamectl 2>/dev/null --printf '%s
%f
%U
%u
%G
%g
%X
%Y
%C')
(stat -L /usr/sbin/hostnamectl 2>/dev/null --printf '%s
%f
%U
%u
%G
%g
%X
%Y
%C')
[2019-04-03T23:03:15-05:00] WARN: Plugin Hostnamectl: did not find hostnamectl
(hostnamectl)
(cat /proc/cmdline || echo -n)
(zpool list -H -o name,size,alloc,free,cap,dedup,health,version)
(test -e /usr/sbin/waagent)
(test -e C:\\WindowsAzure)
(test -e /var/lib/dhcp/dhclient.eth0.leases)
(df -P)
(df -iP)
(mount)
(which lsblk)
(which blkid)
(echo $PATH)
(stat -L /usr/local/bin/blkid 2>/dev/null --printf '%s
%f
%U
%u
%G
%g
%X
%Y
%C')
(stat -L /bin/blkid 2>/dev/null --printf '%s
%f
%U
%u
%G
%g
%X
%Y
%C')
(stat -L /usr/bin'
'/blkid 2>/dev/null --printf '%s
%f
%U
%u
%G
%g
%X
%Y
%C')
(stat -L /usr/bin/blkid 2>/dev/null --printf '%s
%f
%U
%u
%G
%g
%X
%Y
%C')
(stat -L /sbin/blkid 2>/dev/null --printf '%s
%f
%U
%u
%G
%g
%X
%Y
%C')
(/bin/lsblk -n -P -o NAME,UUID,LABEL,FSTYPE)
(test -e /dev/xvda)
(test -e /dev/xvda1)
(/sbin/blkid)
(test -e /proc/mounts)
(cat /proc/mounts || echo -n)
(which getconf)
(/usr/bin/getconf -a)
```

burtlo commented:
Changed the focus of this PR to linux SSH on ec2. Continuing to the next platform I will probably branch off of this one and eventually merge the whole thing.
